File: Python3/401 - 600/434 Number of Segments in a String.py
# 40ms, 71.19%
class Solution:
    def countSegments(self, s):
        if not s: return 0
        count, is_in_segment = 0, False if s[0] == ' ' else True
        for ch in s:
            if is_in_segment:
                if ch == ' ':
                    count += 1
                    is_in_segment = False
            elif ch != ' ':
                is_in_segment = True
        return count if s[-1] == ' ' else count + 1
    

File: 401 - 600/434 Number of Segments in a String.cc
// 3ms, 56.62%
class Solution {
public:
    int countSegments(string s) {
        if (s.empty()) return 0;
        int count = 0;
        bool is_in_segment = s.front() == ' ' ? false : true;
        for (auto ch : s)
            if (is_in_segment) {    // this brace is necessary
                if (ch == ' ') {    // ensure 'else' not mismatch with this 'if'
                    count += 1;
                    is_in_segment = false;
                }
            }
            else if (ch != ' ')
                is_in_segment = true;
        return s.back() == ' ' ? count : count + 1;
    }
};
